simulator/modules/branch/branch.h

```diff
@@ -42,9 +41,20 @@ class Branch : public Log
     public:
         explicit Branch( bool log);
         void clock( Cycle cycle);
-        auto get_mispredict_rate() const { return ( num_branches != 0) 
-                                                  ? num_mispredictions / num_branches * 100
-                                                  : 0; }
+        auto get_mispredictions_num() const { return num_mispredictions; }
+
+        bool is_misprediction( const Instr& instr, const BPInterface& bp_data) const
+        {
+            bool is_misprediction = false;
+
+            is_misprediction |= ( bp_data.is_taken != instr.is_taken()) && !instr.is_likely_branch();
+
+            is_misprediction |= bp_data.is_taken && ( bp_data.target != instr.get_new_PC());
+
+            is_misprediction |= !bp_data.is_taken && !instr.is_taken() && instr.is_likely_branch();
```

Comment:
Once again. What is mispredicted in that case?

Reply:
I figured out 3 situations causing mispredictions which we are able to detect on branch stage.
- BPU predicted to not take branch, but actually we need to take it. And this branch isn't 'likely', otherwise we have caught this misprediction on decode stage.
- BPU predicted to take a jump, but predicted wrong target.
- BPU predicted to not take jump and we actually don't need to take it, but we have taken it because it is 'likely' branch.

Reply:
> And this branch isn't 'likely', otherwise we have caught this misprediction on decode stage.

How?

On decode stage we are able only to check the target. The fact the branch is hinted by compiler does not affect its behavior. The only difference case is case of "cold" branch, i.e., it has no direction prediction. In that case:
1. Common branch is considered to be predicted as "not taken", and next fetched instruction is PC + 4. If it is actually taken, we detect it on B stage and flush the pipeline.
2. Likeliy branch is considered to be predicted as "taken". If the predicted target is not correct (i.e. cold BTB miss), we do a short flush to fetch Target, Target + 4 sequence. If the predicted direction is not correct, we detect it on B stage and flush the pipeline.

Reply:
I thought that one of the ideas of 'likely' branches is that we take a branch and continue with { PC = decoded target}, if BPU decided that this branch **shouldn't be taken**. And we continue with { PC = predicted target}, if BPU **decided to take** it.

So, as I understand now, we should choose the direction **only if BPU said nothing**.
And we should count mispredicts which were made only by BPU.
I will try to fix this.

Reply:
Yes. Likely branches is a instrument useful only in 2 cases:
1. Cold BPU miss.
2. CPU has no BPU at all.

First case is rare in relative to huge amount of BPU hits.
Second case was common for 1980s when MIPS was established, but not now. For instance, RISC-V has no likely branches at all.

Additionally, MIPS likely branches have a difference in processing delay slot, but let's not complicate things now.